### ISSUE TITLE
OrderedDict from typing instantiates to a dict

### ISSUE THREAD

**Issue by saulshanabrook** [pydantic/pydantic#1569]
I want to use a typed ordered dict in Pydantic. I see in the tests that it does support ordereddicts, and this works great.

However, if I use `typing.OrderedDict` instead of `collections.OrderedDict`, so that I can make it generic, it seems to just injest it as a normal dict instead of an ordered one.

Is it currently possible to have a field as an ordered dict with particular types for the keys and values?

Output of `python -c "import pydantic.utils; print(pydantic.utils.version_info())"`:
```
             pydantic version: 1.5.1
            pydantic compiled: True
                 install path: /usr/local/Caskroom/miniconda/base/envs/python-record-api/lib/python3.8/site-packages/pydantic
               python version: 3.8.2 | packaged by conda-forge | (default, Apr 16 2020, 17:20:38)  [Clang 9.0.1 ]
                     platform: macOS-10.15.4-x86_64-i386-64bit
     optional deps. installed: ['typing-extensions']
```

```py
import pydantic
import collections
import typing

class M(pydantic.BaseModel):
    o: typing.OrderedDict[int, typing.Tuple[str, ...]]

# the instance has a dict instead of an ordered dict
assert type(M(o=[]).o) == dict

class M2(pydantic.BaseModel):
    o: collections.OrderedDict

# however if we use collection we get an OrderedDict instance 
assert type(M2(o=[]).o) == collections.OrderedDict
```

p.s. thank you for this library. I was literally just re-creating bits of this myself and grumbling that it was such a PITA in Python, and then I remembered Pydantic exists ❤️

**Comment by saulshanabrook** [comment]
Slightly related https://github.com/samuelcolvin/pydantic/issues/1572

**Comment by adriangb** [comment]
It seems like in V2 we treat both as a `dict`. We have the infrastructure to treat it as an `OrderedDict`, but it doesn't seem to be working, so this is a bug.

**Comment by lig** [comment]
At the moment this cripples test coverage because this https://github.com/pydantic/pydantic/blob/main/pydantic/_internal/_generate_schema.py#L495 is unreachable as we catch all `dict` subclasses here https://github.com/pydantic/pydantic/blob/main/pydantic/_internal/_generate_schema.py#L416.

So, we need to either remove special infrastructure for `OrderedDict` or fix this if we want to see 100% coverage.

**Comment by lig** [comment]
Fixed in #5744
